src/Exceptions/StreamChatException.cs

```diff
@@ -0,0 +1,74 @@
+using System;
+using System.Collections.Generic;
+using System.Net;
+using Newtonsoft.Json;
+using StreamChat.Models;
+using StreamChat.Rest;
+
+namespace StreamChat.Exceptions
+{
+    /// <summary>
+    /// Thrown when the backend returns an unexpected HTTP status code.
+    /// </summary>
+    public class StreamChatException : StreamBaseException
+    {
+        private StreamChatException(ExceptionResponse resp) : base(resp.Message)
+        {
+            ExceptionFields = resp.ExceptionFields;
+            MoreInfo = resp.MoreInfo;
+            ErrorCode = resp.Code;
+            HttpStatusCode = resp.HttpStatusCode;
+            RateLimit = resp.RateLimit;
+        }
+
+        /// <summary>Error code from the backend.</summary>
+        public int ErrorCode { get; }
+
+        /// <summary>HTTP code returned from the backend.</summary>
+        public HttpStatusCode HttpStatusCode { get; }
+
+        /// <summary>Rate limit information returned from the backend.</summary>
+        public RateLimitsInfo RateLimit { get; }
+
+        /// <summary>Exception fields from the backend.</summary>
+        public Dictionary<string, string> ExceptionFields { get; }
+
+        /// <summary>A link that provides more information about the nature of the error.</summary>
+        public Uri MoreInfo { get; }
+
+        internal static StreamChatException FromResponse(RestResponse response)
+        {
+            var resp = new ExceptionResponse { HttpStatusCode = response.StatusCode };
+
+            if (!string.IsNullOrWhiteSpace(response.Content))
+            {
+                resp = JsonConvert.DeserializeObject<ExceptionResponse>(response.Content);
+                resp.HttpStatusCode = response.StatusCode;
+            }
+
+            if (response.TryGetRateLimit(out var rateLimit))
+            {
+                resp.RateLimit = rateLimit;
+            }
+
+            return new StreamChatException(resp);
+        }
+    }
+
+    internal class ExceptionResponse
+    {
+        public int Code { get; set; }
```

Comment:
Code and Message are missing JsonProperty so I think they will not be filled during deserialization

Reply:
not exactly. I only put `JsonProperty` there if the json property name is different. basically when it has underscores. such as `MoreInfo` -> `more_info`. `Message` -> `message` is completely fine by default!